packages/gatsby/src/utils/page-ssr-module/entry.ts

```diff
@@ -0,0 +1,131 @@
+// just types - those should not be bundled
+import type { GraphQLEngine } from "../../schema/graphql-engine/entry"
+import type { IExecutionResult } from "../../query/types"
+import type { IGatsbyPage } from "../../redux/types"
+import type { IScriptsAndStyles } from "../client-assets-for-template"
+import type { IPageDataWithQueryResult } from "../page-data"
+
+// actual imports
+import "../engines-fs-provider"
+import * as path from "path"
+import * as fs from "fs-extra"
+import {
+  constructPageDataString,
+  getPagePathFromPageDataPath,
+} from "../page-data-helpers"
+// @ts-ignore render-page import will become valid later on (it's marked as external)
+import htmlComponentRenderer from "./render-page"
+
+export interface ITemplateDetails {
+  query: string
+  staticQueryHashes: Array<string>
+  assets: IScriptsAndStyles
+}
+export interface ISSRData {
+  results: IExecutionResult
+  page: IGatsbyPage
+  templateDetails: ITemplateDetails
+}
+
+const pageTemplateDetailsMap: Record<
+  string,
+  ITemplateDetails
+  // @ts-ignore INLINED_TEMPLATE_TO_DETAILS is being "inlined" by bundler
+> = INLINED_TEMPLATE_TO_DETAILS
+
+export async function getData({
+  pathName,
+  graphqlEngine,
+}: {
+  graphqlEngine: GraphQLEngine
+  pathName: string
+}): Promise<ISSRData> {
+  const potentialPagePath = getPagePathFromPageDataPath(pathName) || pathName
+
+  // 1. Find a page for pathname
+  const page = graphqlEngine.findPageByPath(potentialPagePath)
+
+  if (!page) {
+    // page not found, nothing to run query for
+    throw new Error(`Page for "${pathName}" not found`)
+  }
+
+  // 2. Lookup query used for a page (template)
+  const templateDetails = pageTemplateDetailsMap[page.componentChunkName]
+  if (!templateDetails) {
+    throw new Error(
+      `Page template details for "${page.componentChunkName}" not found`
+    )
+  }
+
+  // 3. Execute query
+  // query-runner handles case when query is not there - so maybe we should consider using that somehow
+  let results: IExecutionResult = {}
+  if (templateDetails.query) {
+    results = await graphqlEngine.runQuery(templateDetails.query, {
+      ...page,
+      ...page.context,
+    })
+  }
+
+  results.pageContext = page.context
+
+  return { results, page, templateDetails }
+}
+
+export async function renderPageData({
+  data,
+}: {
+  data: ISSRData
+}): Promise<IPageDataWithQueryResult> {
+  const results = await constructPageDataString(
+    {
+      componentChunkName: data.page.componentChunkName,
+      path: data.page.path,
+      matchPath: data.page.matchPath,
+      staticQueryHashes: data.templateDetails.staticQueryHashes,
+    },
+    JSON.stringify(data.results)
+  )
+
+  return JSON.parse(results)
```

Comment:
I assume we have to parse here because of #32763 

Looks a bit weird and we should probably double-check at some point if the benefits of that PR outweigh the overhead here.

Reply:
I think it's possible that we can/will drop `JSON.parse` here and return stringified version of `page-data` instead of json blob. Just I wouldn't want to make those changes in this PR because it's not quite clear to me if it would be safe to do so or not